src/algebra/squarefree.lean

```diff
@@ -204,4 +204,71 @@ lemma sum_divisors_filter_squarefree {n : ℕ} (h0 : n ≠ 0)
 by rw [finset.sum_eq_multiset_sum, divisors_filter_squarefree h0, multiset.map_map,
     finset.sum_eq_multiset_sum]
 
+lemma sq_mul_squarefree {n : ℕ} (hn : 0 < n) :
+  ∃ a b : ℕ, 0 < a ∧ 0 < b ∧ b ^ 2 * a = n ∧ squarefree a :=
+begin
+  let S := {s ∈ finset.range (n + 1) | s ∣ n ∧ ∃ x, s = x ^ 2 },
+  have hSne : S.nonempty,
+  { use 1,
+    have h1 : 0 < n ∧ ∃ (x : ℕ), 1 = x ^ 2 := ⟨hn, ⟨1, (one_pow 2).symm⟩⟩,
+    simpa [S] },
+
+  let s := finset.max' S hSne,
+  have hs : s ∈ S := finset.max'_mem S hSne,
+  simp only [finset.sep_def, S, finset.mem_filter, finset.mem_range] at hs,
+  obtain ⟨hsn1, ⟨a, hsa⟩, ⟨b, hsb⟩⟩ := hs,
+
+  have hzero_lt_a : 0 < a,
+  { rw hsa at hn,
+    rw zero_lt_iff at hn ⊢,
+    exact right_ne_zero_of_mul hn },
+  have hzero_lt_s : 0 < s,
+  { rw hsa at hn,
+    rw zero_lt_iff at hn ⊢,
+    exact left_ne_zero_of_mul hn },
+  have hzero_lt_b : 0 < b,
+  { rw [hsb, pow_two] at hzero_lt_s,
+    rw zero_lt_iff at hzero_lt_s ⊢,
+    exact left_ne_zero_of_mul hzero_lt_s },
+
+  rw hsb at hsa,
```

Comment:
I think the following is an improvement:
```suggestion
  rw hsa at hn,
  obtain ⟨hlts, hlta⟩ := canonically_ordered_semiring.mul_pos.mp hn,
  rw hsb at hsa hn hlts,
  have hltb : 0 < b := by rwa [pow_two, canonically_ordered_semiring.mul_pos, and_self] at hlts,
```
(I shortened the names of some of your assumptions for convenience)